File: src/main/java/akka/first/app/mapreduce/actors/MasterActor.java
package akka.first.app.mapreduce.actors;

import akka.actor.Actor;
import akka.actor.ActorRef;
import akka.actor.Props;
import akka.actor.UntypedActor;
import akka.first.app.mapreduce.messages.MapData;
import akka.first.app.mapreduce.messages.ReduceData;
import akka.first.app.mapreduce.messages.Result;
import akka.japi.Creator;
import akka.routing.RoundRobinPool;
import scala.akka.first.app.mapreduce.actors.ReduceActor;

/**
 * Created by 98379720172 on 06/05/2016.
 */
public class MasterActor extends UntypedActor{

    private static final Creator mapActorCreator = new Creator<Actor>() {
        @Override
        public Actor create() throws Exception {
            return new scala.akka.first.app.mapreduce.actors.MapActor();
        }
    };

    private static final Creator reduceActorCreator = new Creator<Actor>() {
        @Override
        public Actor create() throws Exception {
            return new ReduceActor();
        }
    };

    private static final Creator aggregateActorCreator = new Creator<Actor>() {
        @Override
        public Actor create() throws Exception {
            return new AggregateActor();
        }
    };

    ActorRef mapActor = getContext().actorOf(Props.create(mapActorCreator).withRouter(new RoundRobinPool(5)), "map");
    ActorRef reduceActor = getContext().actorOf(Props.create(reduceActorCreator).withRouter(new RoundRobinPool(5)),"reduce");
    ActorRef aggregateActor = getContext().actorOf(Props.create(aggregateActorCreator), "aggregate");

    public MasterActor() {
    }

    @Override
    public void onReceive(Object message) throws Exception {

        if(message instanceof String){
            mapActor.tell(message, getSelf());
        }else if (message instanceof MapData){
            reduceActor.tell(message, getSelf());
        }else if (message instanceof ReduceData){
            aggregateActor.tell(message, ActorRef.noSender());
        }else if(message instanceof Result){
            aggregateActor.forward(message, getContext());
        }else
            unhandled(message);
    }
}

File: src/main/scala-2.11/scala/akka/first/app/mapreduce/actors/MasterActor.scala
package scala.akka.first.app.mapreduce.actors

/**
  * Created by dsilv on 07/05/2016.
  */

import akka.actor.Actor
import akka.actor.ActorRef
import akka.actor.Props
import akka.first.app.mapreduce._
import akka.routing.RoundRobinPool

import scala.akka.first.app.mapreduce.{MapData, ReduceData, Result}

class MasterActor extends Actor {

  val mapActor = context.actorOf(Props[MapActor].withRouter(
    RoundRobinPool(nrOfInstances = 5)), name = "map")
  val reduceActor = context.actorOf(Props[ReduceActor].withRouter(
    RoundRobinPool(nrOfInstances = 5)), name = "reduce")
  val aggregateActor = context.actorOf(Props[AggregateActor],
    name = "aggregate")

  def receive: Receive = {
    case line: String =>
      mapActor ! line
    case mapData: MapData =>
      reduceActor ! mapData
    case reduceData: ReduceData =>
      aggregateActor ! reduceData
    case Result =>
      aggregateActor forward Result
  }
}
